psbt: Use macro to hash instead of relying on `Hash` trait

tcharding opened the pull request:
We have a private function that makes use of the `Hash` trait to generically hash map entries. This usage makes patching the `hashes` module difficult. We can achieve the same thing by using a macro and passing in the concrete type.

This is an internal change, no effect on logic or public API.

apoelstra reviewed (approved):
ACK 13eb3fbba6dbd709eee0a841c43083e346c127b9

tcharding commented:
Changed PR name to make searching easier because this is related to the `hashes` crate.

tcharding commented:
Changes in force push are fixes to my sloppy programming:

- Remove "delete sha256t module" - don't know how that snuck in there.
- Rebased on master
- Removed code comment that was just for me to see how the macro was called during dev

apoelstra commented:
> * Remove "delete sha256t module" - don't know how that snuck in there.

Unsure what this refers to. I can't find any such thing in this version of the PR or the previous version.

But thanks for dropping the errant comment and tweaking whitespace :).

apoelstra reviewed (approved):
ACK 9e4b092fce79501cb8cdd650fc34f66d01db7a7b

tcharding commented:
> Unsure what this refers to. 

Ugh, I literally thought while writing that "I wonder if Andrew will see the red diff lines that this refers to now I've force pushed" - and wrote it anyways :(

tcharding commented:
Gentle bump.

apoelstra commented:
Heads up that the API CI job is broken, presumably due to implicit merge conflicts. I'll keep merging stuff and then try to update my local CI to start checking this (it's a bit hard to add new cargo tools with tons of deps).
